src/upnp.cpp

```diff
@@ -225,7 +225,7 @@ int upnp::add_mapping(portmap_protocol const p, int const external_port
 		m.external_port = external_port;
 		m.local_port = local_port;
 
-		if (d.service_namespace) update_map(d, mapping_index);
+		if (!d.service_namespace.empty()) update_map(d, mapping_index);
```

Comment:
why not

`if (d.service_namespace && !d.service_namespace.empty()) update_map(d, mapping_index);`

and avoid a possible NPE?

The same comment applies everywhere else below, and since it happens so much perhaps I'd add a private utility function to not repeat yourself all over

```
bool valid_service_namespace(rootdevice &d) {
    return d.service_namespace && !d.service_namespace.empty();
}
```

and then have

```
if (valid_service_namespace(d)) update_map(d, mapping_index);

...

TORRENT_ASSERT(valid_service_namespace(d));

...
```